File: reference/src/policy/enforcement.ts
/**
 * Quality gate enforcement engine.
 * Implements the 3-tier enforcement model from spec/policy.md.
 */

import type {
  Gate,
  EnforcementLevel,
  QualityGate,
  ToolRule,
  ReviewerRule,
  DocumentationRule,
  ProvenanceRule,
  CostRule,
  DesignTokenComplianceRule,
  VisualRegressionRule,
  StoryCompletenessRule,
  DesignReviewGateRule,
  DesignReviewDecision,
} from '../core/types.js';
import { compareMetric, exceedsSeverity } from '../core/compare.js';

export interface EvaluationContext {
  authorType: 'ai-agent' | 'human' | 'bot' | 'service-account';
  repository: string;
  metrics: Record<string, number>;
  overrideRole?: string;
  overrideJustification?: string;
  toolResults?: Record<
    string,
    { findings: { severity: 'low' | 'medium' | 'high' | 'critical' }[] }
  >;
  reviewerCount?: number;
  changedFiles?: string[];
  docFiles?: string[];
  provenance?: { attribution?: boolean; humanReviewed?: boolean };
  /** Design review context for designReview gate rules (RFC-0006 §8.5). */
  designReview?: {
    decision?: DesignReviewDecision;
    reviewer?: string;
    pendingTimeout?: boolean;
  };
  /** Design token compliance violations by category (RFC-0006 §8.1). */
  designTokenViolations?: Record<string, number>;
  /** Story metadata for completeness checks (RFC-0006 §8.3). */
  storyMeta?: {
    hasDefaultStory?: boolean;
    hasStateStories?: boolean;
    hasA11yStory?: boolean;
    storyCount?: number;
  };
}

export type GateVerdict = 'pass' | 'fail' | 'override';

export interface GateResult {
  gate: string;
  enforcement: EnforcementLevel;
  verdict: GateVerdict;
  message?: string;
}

export interface EnforcementResult {
  allowed: boolean;
  results: GateResult[];
}

interface RuleResult {
  passed: boolean;
  message?: string;
}

/**
 * Evaluate a single rule against the provided context.
 */
function evaluateRule(rule: Gate['rule'], ctx: EvaluationContext): RuleResult {
  // Metric-based rule
  if ('metric' in rule && 'operator' in rule && 'threshold' in rule) {
    const actual = ctx.metrics[rule.metric];
    if (actual === undefined) {
      return { passed: false, message: `Metric "${rule.metric}" not available` };
    }
    return {
      passed: compareMetric(actual, rule.operator as string, rule.threshold as number),
    };
  }

  // Tool-based rule
  if ('tool' in rule) {
    const toolRule = rule as ToolRule;
    const results = ctx.toolResults?.[toolRule.tool];
    if (!results) {
      return { passed: false, message: `Tool "${toolRule.tool}" results not available` };
    }
    if (toolRule.maxSeverity) {
      const violations = results.findings.filter((f) =>
        exceedsSeverity(f.severity, toolRule.maxSeverity!),
      );
      if (violations.length > 0) {
        return {
          passed: false,
          message: `${violations.length} finding(s) exceed max severity "${toolRule.maxSeverity}"`,
        };
      }
    }
    return { passed: true };
  }

  // Reviewer-based rule
  if ('minimumReviewers' in rule) {
    const reviewerRule = rule as ReviewerRule;
    let required = reviewerRule.minimumReviewers;
    if (reviewerRule.aiAuthorRequiresExtraReviewer && ctx.authorType === 'ai-agent') {
      required += 1;
    }
    const actual = ctx.reviewerCount ?? 0;
    if (actual >= required) {
      return { passed: true };
    }
    return {
      passed: false,
      message: `Requires ${required} reviewer(s), got ${actual}`,
    };
  }

  // Documentation-based rule
  if ('changedFilesRequireDocUpdate' in rule) {
    const docRule = rule as DocumentationRule;
    if (!docRule.changedFilesRequireDocUpdate) {
      return { passed: true };
    }
    const hasCodeChanges = (ctx.changedFiles ?? []).length > 0;
    const hasDocChanges = (ctx.docFiles ?? []).length > 0;
    if (hasCodeChanges && !hasDocChanges) {
      return { passed: false, message: 'Code changes require documentation updates' };
    }
    return { passed: true };
  }

  // Cost-based rule
  if ('cost' in rule) {
    const costRule = rule as CostRule;
    const actual = ctx.metrics[costRule.cost.metric];
    if (actual === undefined) {
      return { passed: false, message: `Cost metric "${costRule.cost.metric}" not available` };
    }
    return {
      passed: compareMetric(actual, costRule.cost.operator, costRule.cost.threshold),
    };
  }

  // Provenance-based rule
  if ('requireAttribution' in rule) {
    const provRule = rule as ProvenanceRule;
    const prov = ctx.provenance ?? {};
    if (provRule.requireAttribution && !prov.attribution) {
      return { passed: false, message: 'Attribution is required' };
    }
    if (provRule.requireHumanReview && !prov.humanReviewed) {
      return { passed: false, message: 'Human review is required' };
    }
    return { passed: true };
  }

  // Design token compliance rule (RFC-0006 §8.1)
  if ('designTokenCompliance' in rule) {
    const dtcRule = rule as DesignTokenComplianceRule;

    // Coverage metric mode
    if (dtcRule.coverageMetric) {
      const coverage = ctx.metrics['token-coverage'];
      if (coverage === undefined) {
        return { passed: false, message: 'Metric "token-coverage" not available' };
      }
      return {
        passed: compareMetric(
          coverage,
          dtcRule.coverageMetric.operator,
          dtcRule.coverageMetric.threshold,
        ),
        message: !compareMetric(
          coverage,
          dtcRule.coverageMetric.operator,
          dtcRule.coverageMetric.threshold,
        )
          ? `Token coverage ${coverage} does not meet ${dtcRule.coverageMetric.operator} ${dtcRule.coverageMetric.threshold}`
          : undefined,
      };
    }

    // Violation count mode
    const category = dtcRule.category ?? 'all';
    const violations =
      ctx.designTokenViolations?.[category] ?? ctx.metrics[`${category}-violations`] ?? 0;
    const max = dtcRule.maxViolations ?? 0;
    if (violations > max) {
      return {
        passed: false,
        message: `${violations} ${category} token violation(s) exceed max ${max}`,
      };
    }
    return { passed: true };
  }

  // Visual regression rule (RFC-0006 §8.2)
  if ('visualRegression' in rule) {
    const vrRule = rule as VisualRegressionRule;
    const diffPct = ctx.metrics['visual-diff-percentage'];

    if (vrRule.config.requireBaseline && ctx.metrics['baseline-exists'] === 0) {
      return { passed: false, message: 'Visual baseline does not exist' };
    }

    if (diffPct === undefined) {
      return { passed: false, message: 'Metric "visual-diff-percentage" not available' };
    }

    if (diffPct > vrRule.config.diffThreshold) {
      return {
        passed: false,
        message: `Visual diff ${(diffPct * 100).toFixed(1)}% exceeds threshold ${(vrRule.config.diffThreshold * 100).toFixed(1)}%`,
      };
    }
    return { passed: true };
  }

  // Story completeness rule (RFC-0006 §8.3)
  if ('storyCompleteness' in rule) {
    const scRule = rule as StoryCompletenessRule;
    const meta = ctx.storyMeta;
    if (!meta) {
      return { passed: false, message: 'Story metadata not available' };
    }

    if (scRule.config.requireDefaultStory && !meta.hasDefaultStory) {
      return { passed: false, message: 'Default story is required but missing' };
    }
    if (scRule.config.requireStateStories && !meta.hasStateStories) {
      return { passed: false, message: 'Interactive state stories are required but missing' };
    }
    if (scRule.config.requireA11yStory && !meta.hasA11yStory) {
      return { passed: false, message: 'Accessibility story is required but missing' };
    }
    if (scRule.config.minStories && (meta.storyCount ?? 0) < scRule.config.minStories) {
      return {
        passed: false,
        message: `Component has ${meta.storyCount ?? 0} stories, minimum ${scRule.config.minStories} required`,
      };
    }
    return { passed: true };
  }

  // Design review rule (RFC-0006 §8.5)
  if ('designReview' in rule) {
    const drRule = rule as DesignReviewGateRule;
    const review = ctx.designReview;

    if (!review) {
      return { passed: false, message: 'Design review not yet submitted' };
    }

    if (review.pendingTimeout) {
      if (drRule.onTimeout === 'fail') {
        return { passed: false, message: 'Design review timed out' };
      }
      // 'pause' means the pipeline should hold — treat as not-yet-passed
      return { passed: false, message: 'Design review pending (timed out, paused)' };
    }

    if (review.decision === 'approved' || review.decision === 'approved-with-comments') {
      return { passed: true };
    }

    return {
      passed: false,
      message: `Design review ${review.decision ?? 'pending'} by ${review.reviewer ?? 'unknown'}`,
    };
  }

  return { passed: false, message: 'Unknown rule type' };
}

/**
 * Evaluate a single gate against the provided context.
 */
export function evaluateGate(gate: Gate, ctx: EvaluationContext): GateResult {
  const result = evaluateRule(gate.rule, ctx);

  if (result.passed) {
    return { gate: gate.name, enforcement: gate.enforcement, verdict: 'pass' };
  }

  // Check for soft-mandatory override (applies to all rule types)
  if (gate.enforcement === 'soft-mandatory' && gate.override && ctx.overrideRole) {
    if (ctx.overrideRole === gate.override.requiredRole) {
      if (!gate.override.requiresJustification || ctx.overrideJustification) {
        return {
          gate: gate.name,
          enforcement: gate.enforcement,
          verdict: 'override',
          message: `Overridden by ${ctx.overrideRole}`,
        };
      }
    }
  }

  return {
    gate: gate.name,
    enforcement: gate.enforcement,
    verdict: 'fail',
    message: result.message,
  };
}

/**
 * Evaluate all gates in a QualityGate resource and determine whether
 * the action is allowed.
 *
 * Enforcement semantics:
 * - advisory: logged but never blocks
 * - soft-mandatory: blocks unless overridden by authorized role
 * - hard-mandatory: always blocks on failure, no override
 */
export function enforce(qualityGate: QualityGate, ctx: EvaluationContext): EnforcementResult {
  const results = qualityGate.spec.gates.map((gate) => evaluateGate(gate, ctx));

  const allowed = results.every((r) => {
    if (r.verdict === 'pass' || r.verdict === 'override') return true;
    if (r.enforcement === 'advisory') return true;
    return false;
  });

  return { allowed, results };
}

File: sdk-python/src/ai_sdlc/policy/enforcement.py
"""Quality gate enforcement engine.

Implements the 3-tier enforcement model from spec/policy.md.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import TYPE_CHECKING, Any, Literal

from ai_sdlc.core.compare import compare_metric, exceeds_severity

if TYPE_CHECKING:
    from ai_sdlc.core.types import (
        EnforcementLevel,
        Gate,
        QualityGate,
    )

from ai_sdlc.core.types import (
    DocumentationRule,
    ExpressionRule,
    MetricRule,
    ProvenanceRule,
    ReviewerRule,
    ToolRule,
)


@dataclass(frozen=True)
class EvaluationContext:
    author_type: Literal["ai-agent", "human", "bot", "service-account"]
    repository: str
    metrics: dict[str, float]
    override_role: str | None = None
    override_justification: str | None = None
    tool_results: dict[str, dict[str, Any]] | None = None
    reviewer_count: int | None = None
    changed_files: list[str] | None = None
    doc_files: list[str] | None = None
    provenance: dict[str, bool] | None = None


GateVerdict = Literal["pass", "fail", "override"]


@dataclass(frozen=True)
class GateResult:
    gate: str
    enforcement: EnforcementLevel
    verdict: GateVerdict
    message: str | None = None


@dataclass(frozen=True)
class EnforcementResult:
    allowed: bool
    results: list[GateResult] = field(default_factory=list)


@dataclass(frozen=True)
class _RuleResult:
    passed: bool
    message: str | None = None


def _evaluate_rule(rule: Gate, ctx: EvaluationContext) -> _RuleResult:
    """Evaluate a single rule against the provided context."""
    r = rule.rule

    # Metric-based rule
    if isinstance(r, MetricRule):
        actual = ctx.metrics.get(r.metric)
        if actual is None:
            return _RuleResult(
                passed=False,
                message=f'Metric "{r.metric}" not available',
            )
        return _RuleResult(
            passed=compare_metric(actual, r.operator, r.threshold),
        )

    # Tool-based rule
    if isinstance(r, ToolRule):
        results = (ctx.tool_results or {}).get(r.tool)
        if results is None:
            return _RuleResult(
                passed=False,
                message=f'Tool "{r.tool}" results not available',
            )
        if r.max_severity:
            findings = results.get("findings", [])
            violations = [
                f for f in findings
                if exceeds_severity(f["severity"], r.max_severity)
            ]
            if violations:
                return _RuleResult(
                    passed=False,
                    message=(
                        f'{len(violations)} finding(s) exceed'
                        f' max severity "{r.max_severity}"'
                    ),
                )
        return _RuleResult(passed=True)

    # Reviewer-based rule
    if isinstance(r, ReviewerRule):
        required = r.minimum_reviewers
        if r.ai_author_requires_extra_reviewer and ctx.author_type == "ai-agent":
            required += 1
        actual_count = ctx.reviewer_count or 0
        if actual_count >= required:
            return _RuleResult(passed=True)
        return _RuleResult(
            passed=False,
            message=f"Requires {required} reviewer(s), got {actual_count}",
        )

    # Documentation-based rule
    if isinstance(r, DocumentationRule):
        if not r.changed_files_require_doc_update:
            return _RuleResult(passed=True)
        has_code_changes = len(ctx.changed_files or []) > 0
        has_doc_changes = len(ctx.doc_files or []) > 0
        if has_code_changes and not has_doc_changes:
            return _RuleResult(
                passed=False,
                message="Code changes require documentation updates",
            )
        return _RuleResult(passed=True)

    # Provenance-based rule
    if isinstance(r, ProvenanceRule):
        prov = ctx.provenance or {}
        if r.require_attribution and not prov.get("attribution"):
            return _RuleResult(passed=False, message="Attribution is required")
        if r.require_human_review and not prov.get("humanReviewed"):
            return _RuleResult(
                passed=False, message="Human review is required",
            )
        return _RuleResult(passed=True)

    # ExpressionRule — not evaluated directly in enforcement
    if isinstance(r, ExpressionRule):
        return _RuleResult(passed=False, message="Expression rules not yet supported")

    return _RuleResult(passed=False, message="Unknown rule type")


def evaluate_gate(gate: Gate, ctx: EvaluationContext) -> GateResult:
    """Evaluate a single gate against the provided context."""
    result = _evaluate_rule(gate, ctx)

    if result.passed:
        return GateResult(gate=gate.name, enforcement=gate.enforcement, verdict="pass")

    # Check for soft-mandatory override
    if (
        gate.enforcement == "soft-mandatory"
        and gate.override
        and ctx.override_role
        and ctx.override_role == gate.override.required_role
        and (not gate.override.requires_justification or ctx.override_justification)
    ):
        return GateResult(
            gate=gate.name,
            enforcement=gate.enforcement,
            verdict="override",
            message=f"Overridden by {ctx.override_role}",
        )

    return GateResult(
        gate=gate.name,
        enforcement=gate.enforcement,
        verdict="fail",
        message=result.message,
    )


def enforce(quality_gate: QualityGate, ctx: EvaluationContext) -> EnforcementResult:
    """Evaluate all gates in a QualityGate resource.

    Enforcement semantics:
    - advisory: logged but never blocks
    - soft-mandatory: blocks unless overridden by authorized role
    - hard-mandatory: always blocks on failure, no override
    """
    results = [evaluate_gate(g, ctx) for g in quality_gate.spec.gates]

    allowed = all(
        r.verdict in ("pass", "override") or r.enforcement == "advisory"
        for r in results
    )

    return EnforcementResult(allowed=allowed, results=results)
